### ISSUE TITLE
7 Series Power state not recognized

### ISSUE THREAD

**Issue by jnnksdev** [kdschlosser/samsungctl#132]
The code is stuck in a loop to wait for the TV to power on, but it's already running. The [property change](https://github.com/kdschlosser/samsungctl/blob/90756acc7ac210e37bbfa23dccf1ca3113e55154/samsungctl/remote_websocket.py#L61)  never occurrs.

The event loop does not seem to recognize the Tvs response.

```
# stuck at ctor
with samsungctl.Remote(config) as remote:
    print(remote.has_ssl())
```
```
GET /api/v2/ HTTP/1.1
Host: 192.168.0.66:8001
User-Agent: python-requests/2.22.0
Accept-Encoding: gzip, deflate
Accept: */*
Connection: keep-alive
```

```
HTTP/1.1 200 OK
content-type: application/json; charset=utf-8
content-length: 1315

"device":{
    "FrameTVSupport":"false",
    "GamePadSupport":"true",
    "ImeSyncedSupport":"true",
    "Language":"en_GB",
    "OS":"Tizen",
    "PowerState":"on",
    "TokenAuthSupport":"true",
    "VoiceSupport":"false",
    "WallScreenRatio":"0",
    "WallService":"false",
    "countryCode":"DE",
    "description":"Samsung DTV RCR",
    "developerIP":"0.0.0.0",
    "developerMode":"1",
    "duid":"uuid:x",
    "firmwareVersion":"Unknown",
    "id":"uuid:x",
    "ip":"192.168.0.66",
    "model":"19_MUSEL_UHD_BASIC",
    "modelName":"UE55RU7099UXZG",
    "name":"[TV] Samsung 7 Series (55)",
    "networkType":"wireless",
    "resolution":"3840x2160",
    "smartHubAgreement":"true",
    "ssid":"ac:22:05:e4:67:9b",
    "type":"Samsung SmartTV",
    "udn":"uuid:x",
    "wifiMac":"D0:D0:03:7E:2E:A1"
},
   "id":"uuid:x",
   "isSupport":"{\"DMP_DRM_PLAYREADY\":\"false\",\"DMP_DRM_WIDEVINE\":\"false\",\"DMP_available\":\"true\",\"EDEN_available\":\"true\",\"FrameTVSupport\":\"false\",\"ImeSyncedSupport\":\"true\",\"TokenAuthSupport\":\"true\",\"remote_available\":\"true\",\"remote_fourDirections\":\"true\",\"remote_touchPad\":\"true\",\"remote_voiceControl\":\"false\"}\n",
   "name":"[TV] Samsung 7 Series (55)",
   "remote":"1.0",
   "type":"Samsung SmartTV",
   "uri":"http://192.168.0.66:8001/api/v2/",
   "version":"2.0.25"
```

my config:
```
    "host": "192.168.0.66",
    "port": 8001,
    "method": "websocket",
    "name": "r6",
    "description": "",
    "id": "",
    "token": "",
    "timeout": 1
```
The timeout settings does not seem to make a difference, as does method encrypted/websocket.

log:
```
RemoteWrapper.open()
RemoteWrapper.power()
RemoteWrapper.power => False
RemoteWrapper.power(value=True)
RemoteWrapper.power()
RemoteWrapper.power => False
RemoteWrapper.mac_address()
RemoteWrapper.mac_address => 'D0:D0:03:7E:2E:A1'
RemoteWrapper.mac_address()
RemoteWrapper.mac_address => 'D0:D0:03:7E:2E:A1'
RemoteWrapper.power()
RemoteWrapper.power => False
RemoteWrapper.open()
RemoteWrapper.mac_address()
RemoteWrapper.mac_address => 'D0:D0:03:7E:2E:A1'
RemoteWrapper.power()
RemoteWrapper.power => False
RemoteWrapper.open()
```

**Comment by jnnksdev** [comment]
It's a deadlock because of recursive method calls.

[open()](https://github.com/kdschlosser/samsungctl/blob/90756acc7ac210e37bbfa23dccf1ca3113e55154/samsungctl/remote_websocket.py#L52)
    [lock](https://github.com/kdschlosser/samsungctl/blob/90756acc7ac210e37bbfa23dccf1ca3113e55154/samsungctl/remote_websocket.py#L57)
    [power on](https://github.com/kdschlosser/samsungctl/blob/90756acc7ac210e37bbfa23dccf1ca3113e55154/samsungctl/remote_websocket.py#L61) -> [power()](https://github.com/kdschlosser/samsungctl/blob/90756acc7ac210e37bbfa23dccf1ca3113e55154/samsungctl/remote_websocket.py#L214)
[open() #2](https://github.com/kdschlosser/samsungctl/blob/90756acc7ac210e37bbfa23dccf1ca3113e55154/samsungctl/remote_websocket.py#L226)

**Comment by xMarkos** [comment]
The problem is still present as of today. While inspecting the code, I realized that the offending recursion to the open() method can be prevented by using configuration setting paired=True.
This workaround works even if the TV is not paired yet, at least with my TV model.
Hope it saves somebody some nerves.

**Comment by avg-I** [comment]
Yeah, the problem is quite annoying. Especially because it's so hard to get through various indirections in the code.
Like that `power` attribute actually means whether the websocket connection is made (so, has very little to do with the actual power state).
And setting `power` sends a WOL packet and then just sleeps and then recurses to `open`.

**Comment by ThorstenF77** [comment]
I have the same TV. The problem is that it expect a TLS connection with a self-signed certificate and a token that needs to be passed after the first time the remote was permitted. The following code works for me:

`import base64
import json
import logging
import time
import websocket
import ssl
from websocket import create_connection

class SamsungTV():
	def __init__(self, host, port = 8002, name = "MyRemote", token = None):
		uri = "wss://" + str(host) + ":"+ str(port) + "/api/v2/channels/samsung.remote.control?name=" + base64.b64encode(str.encode(name)).decode('utf-8')
		if token != None:
			uri += "&token=" + str(token)
		self.connection = websocket.create_connection(uri , sslopt={"cert_reqs": ssl.CERT_NONE})

		response = json.loads(self.connection.recv())
		# print(response)
		if response['event'] != 'ms.channel.connect':
			self.close()
			raise Exception(response)
		self.token = str(response['data']['clients'][0]['attributes']['token'])

	def __exit__(self, type, value, traceback):
		self.close()

	def close(self):
		if self.connection:
			self.connection.close()
			self.connection = None

	def get_token(self):
		return self.token

	def send_key(self, key, repeat=1):
		for n in range(repeat):
			payload = json.dumps({'method': 'ms.remote.control','params': {'Cmd': 'Click','DataOfCmd': key,'Option': 'false','TypeOfRemote': 'SendRemoteKey'}})
			self.connection.send(payload)
			# time.sleep(1)
		
tv = SamsungTV("INSERT-IP-ADDRESS", token=None) # of course your IP, and leave token blank on first run, second run put the string in that it printed
print("Token: " + tv.get_token())
tv.send_key("KEY_HOME")
tv.send_key("KEY_INFO")
`

**Comment by kdschlosser** [comment]
This is a very old library that has not been maintained in a long time. The reason why it has not been maintained is because Samsung sucks and they were changing the API and breaking things left and right. At one point in time their own application would not even pair to the TV's because of changes they made to the firmware for the TV's. Frankly I am kind of surprised it is even working at all.

The reason why things are coded how they are is because there is no real way of being able to collect the power state from the TV's and because there are several different types of commands that have to be sent to the TV depending on the year the TV was made.

**Comment by anohren** [comment]
After working around the deadlock to get a config file this seems to work nicely with an NU-model
